scene/3d/sprite_3d.cpp

```diff
@@ -851,7 +877,20 @@ void AnimatedSprite3D::_draw() {
 		}
 	}
 	set_aabb(aabb);
-	VS::get_singleton()->immediate_end(immediate);
+
+	Array mesh_array;
+	mesh_array.resize(VS::ARRAY_MAX);
+	mesh_array[VS::ARRAY_VERTEX] = mesh_vertices;
+	mesh_array[VS::ARRAY_NORMAL] = mesh_normals;
+	mesh_array[VS::ARRAY_TANGENT] = mesh_tangents;
+	mesh_array[VS::ARRAY_COLOR] = mesh_colors;
+	mesh_array[VS::ARRAY_TEX_UV] = mesh_uvs;
+
+	VS::get_singleton()->mesh_add_surface_from_arrays(mesh, VS::PRIMITIVE_TRIANGLE_FAN, mesh_array);
+	RID mat = SpatialMaterial::get_material_rid_for_2d(get_draw_flag(FLAG_SHADED), get_draw_flag(FLAG_TRANSPARENT), get_draw_flag(FLAG_DOUBLE_SIDED), get_alpha_cut_mode() == ALPHA_CUT_DISCARD, get_alpha_cut_mode() == ALPHA_CUT_OPAQUE_PREPASS, get_billboard_mode() == SpatialMaterial::BILLBOARD_ENABLED, get_billboard_mode() == SpatialMaterial::BILLBOARD_FIXED_Y);
+	VS::get_singleton()->material_set_shader(get_material(), VS::get_singleton()->material_get_shader(mat));
+	VS::get_singleton()->material_set_param(get_material(), "texture_albedo", texture->get_rid());
+	VS::get_singleton()->instance_set_surface_material(get_instance(), 0, get_material());
```

Comment:
If _draw is being called every frame for the AnimatedSprite, it may be worth doing a naive check for whether these states have changed from this side, depending on how involved the checks are for a null change in the visual server.

I.e. If texture->get_rid() == last rid, don't set in visual server.

Reply:
I think this is a very good idea. But I don't think it is necessary. ``_draw()`` is only called when an update has been queued. Updates are queued when a property changes that would affect the appearance of the Sprite3D.  

As you pointed out, the only real issue is for AnimatedSprites because ``_draw()`` is called every frame. But the reason ``_draw()`` is called every frame is because the texture is changing every frame (or the texture region or something similar). The only real opportunity for optimization is the case where a texture atlas is used and every frame is from the same atlas, then we could potentially save on the rebind. However, the cost is so low, I don't think the added complexity is worth it unless we find out that it is creating a bottleneck.